### ISSUE TITLE
Permit to easily switch from one device (eg PC) to another one (eg android)

### ISSUE THREAD

**Issue by Kcchouette** [simplex-chat/simplex-chat#444]
Hello

I didn't find the way to easily switch from one device (eg PC) to another one (eg android), so I'm opening this issue to permit that.

Thanks in advance

**Comment by epoberezkin** [comment]
You have to copy the files, that’s currently only possible with rooted Android device I think - these are internal app files.

we will add the ability to import / export the database files, it is definitely coming.

if you were to copy files now, please make sure to always use the latest backup and only from one device at a time.

**Comment by cmroanirgo** [comment]
I hope i'm on the same page as the original commenter, but i'm relatively sure import/export isn't what was envisaged. I'm more thinking of a seamless swapping, like IMAP does for email: read on any device & it's read everywhere/ send on any device & it's seen as sent on other devices. 

Obviously, there would need to be inter-device permissions for this to be allowed, including device revoke (of any permissioned device by any permissioned device).

**Comment by epoberezkin** [comment]
Export / import was just added btw.

I understand - proper profile sync is coming some day. It's not a very high priority as it is quite complex... Groups are coming sooner, and groups is another clutch that can serve as a workaround for missing profile sync for many users - you can have groups instead of 121 chats, just adding all devices of two people in the group - we use it for a long time already.

**Comment by zw7u9tru** [comment]
Just wanted to say that this feature is important to me from an accessibility standpoint (poor/declining vision). Typing on a keyboard is much easier, but not always available.


 It's ok if behind the scenes groups are used, but an easy way to switch from my phone to my computer without having to turn every chat into a group chat and having to add both devices every time would make me significantly more likely to use this app.

**Comment by trymeouteh** [comment]
Please allow for multiple device syncing

- If a message is being sent to you, it will appear on all your devices
- Messages you send will appear on all of your devices
- Incoming calls will ring on all of your devices and the device you use to answer the call will be the device the call is on

**Comment by xxxserxxx** [comment]
Multi-device sync could be implemented as a shim over groups, with only an addition to the spec (to ensure consistent user experience) and a little bit of client hand-waving. 

It'd require four changes: 

1. allowing multiple keys to refer to the same person 
2. a self-invite linking devices and at the same time approving the other device as being "also me" 
3. clients validating when multiple entities are the same person (this may be the bulk of the work)
4. clients displaying all messages from keys under the same name

This fails in some obvious ways: without more work, history editing wouldn't work for cross-device messages, for instance. I'm sure there are others. But as mentioned, proper cross-device synchronization and multiple-device account support can get complex, when most of the functionality could be hidden on the back of groups with relatively little additional effort. There would essentially be no device management, such as currents plagues Matrix. Yes, there's membership management, but that's already a necessity under group support.

**Comment by Alch-Emi** [comment]
One other thing that might be worth considering is data usage.  Since messages are padded to 16KiB and all SMP channels are 1:1, sending a message to multiple users can start to use a lot of data.  Imagine a group chat with 10 users, each with four devices.  Sending a single message in this room would use 639 KB of data, assuming the message is sent to every device of every user, except the sender device.  Sending four messages brings this to 2.5MB.

I don't think this scenario is much of an edge case, especially considering that it's easy for old devices to accumulate (see: matrix), and 10 people isn't an unreasonable group chat size.  But 2.5MB is also a *lot* of data for four short messages (e.g. "got the address", "im omw", "cya in 10m", "<3"), especially if you're on a metered connection or have poor signal, which many mobile users are likely to be.

I'm not really sure what the solution to this is, but I think it should be at least a consideration.

(footnote:  I might be misunderstanding the SMP protocol and how SimpleX handles groups on top of that — please correct me if I've misunderstood)

**Comment by epoberezkin** [comment]
Yes, your calculations are correct. We are working on the ability to use mobile profiles on desktop.

**Comment by vecna13** [comment]
Would you consider making device linking independent of the type of device? I'd like to use a desktop device as my primary device and link an Android device to it. (My use case is having an extra Android device I don't carry around but just use while at home and lying down due to its smaller form factor. I'd like to use it occasionally but not as the primary device.)

**Comment by Mikrizi** [comment]
+, it's very need

> Hello
> 
> I didn't find the way to easily switch from one device (eg PC) to another one (eg android), so I'm opening this issue to permit that.
> 
> Thanks in advance

**Comment by trymeouteh** [comment]
Please remove the "master device" and "children devices" setup.

I would prefer if any device is a "master device" which will make it more reliable for the user if one of their devices is lost or broken.

And to have this syncing between devices done per profile, not all profiles on a device. This way a users phone can have their personal profile while the users laptop can have their personal and work profile, and not have their work profile on their phone if they so wish.

**Comment by epoberezkin** [comment]
https://x.com/SimpleXChat/status/1797731573711695913

**Comment by mak448a** [comment]
> https://x.com/SimpleXChat/status/1797731573711695913

Does this mean that device syncing is not planned?

**Comment by jermanuts** [comment]
@mak448a it is planned, but they are taking their time to implement a one that is secure.

**Comment by erebion** [comment]
Please... "master device" is just racist terminology. At least talk about "primary devices". 🙄

Also, unless there's a very good reason to have a primary device, I'd much rather prefer not having to make sure one device never breaks to not lose access.

**Comment by Yiannis128** [comment]
I agree that this is the only required thing left before I fully switch to SimpleX. Feeds would be nice too (I can see that they're planned) as it would allow me to switch from Telegram, which does not use a standardized encryption protocol and is also centralized.

> Please... "master device" is just racist terminology. At least talk about "primary devices". 🙄
> 
> Also, unless there's a very good reason to have a primary device, I'd much rather prefer not having to make sure one device never breaks to not lose access.

The thing is, not having a master device makes syncronization much more complex. Also, the benefit of having a master device is that you can have that as the master authority over all other devices. So the master device could be used to revoke access if another device is compromized. Without a master device, if someone gains access to one of your devices, they can't be kicked out since they will have access to your keys.

I do agree that there is a disadvantage of having to have the master device always sync first to communicate with the other ones, but maybe there probably are workarounds. I am not familiar with the SimpleX algorithm so I can't say.

**Comment by ElsAr4e** [comment]
@epoberezkin Will the new "real"  multi-device also be as secure as the new Threema multi-device functionality? https://threema.ch/en/blog/posts/md-architectural-overview

Also Threema claims that the way they've implemented multi-device support has been done like no other messenger (German website; you can use a translator) https://threema-forum.de/index.php?thread/7047-threema-desktop-2-0/&postID=76870#post76870

**Comment by m6n50u7** [comment]
this doesn't need any server side edition.
all you have to do is to add another hidden "chat" in the client side to do the sync job without having to pair the different clients/devices more than the first time.

**Comment by Mart-Bogdan** [comment]
I've tried importing profile backed up on desktop onto mobile devices and I've got pretty broken set-up

I was able to see previous messages, but new messages seems to work only on one device. In general set up become quite broken.

But I was able to create two 1 on 1 chats with myself. same name and user pic but different chats.  It's pretty broken and confusing.

As I see -- account recovery of data not being backed up isnt possible due to privacy.

**Comment by SamiLehtinen** [comment]
> But I was able to create two 1 on 1 chats with myself. same name and user pic but different chats. It's pretty broken and confusing.

Sounds like a totally expected outcome. I don't see anything broken there.
 
> As I see -- account recovery of data not being backed up isnt possible due to privacy.

There are no accounts to recover in the first place, so account recovery can't be broken.

**Comment by mjobin-mdsol** [comment]
@m6n50u7 I like your suggestion... it could sync in the background with a trusted list of authorized devices
I just wonder which devices gets the message first can be random and the message ordering could be messed up.
especially for group chat, not so much for 1-1 I guess

**Comment by Katarn** [comment]
Synchronization logic from Resilio Sync/Syncthing is not suitable? No master device is used.

**Comment by Mathias-g** [comment]
Please ignore if stupid, I don't currently understand Simplex deeply enough to say it would work, but couldn't you just serialize messages on the SMP server before they are sent out to the client and then it doesn't matter which order they arrive in as the client can just order them once they arrive?

**Comment by mjobin-mdsol** [comment]
@Katarn The way I see it, without as primary device, there is a risk of miss-ordering between the sender and receiver... but even that is probably solvable... I don't see how the primary device is required, you got a list of messages on both sides, some are missing, they should sync, so everything is there.

**Comment by mjobin-mdsol** [comment]
@Mathias-g I am not an expert on the internals either, I barely started reading on this as I really care about being able to use two Android phones and multiple Linux computers.

But, like you said, I think it would be possible to order them as they hit the server, I think the main worries is we don't want timestamps to be part of the metadata. Reading [the protocol doc here](https://github.com/simplex-chat/simplexmq/blob/stable/protocol/simplex-messaging.md), I can't find if its part of the metadata held on servers.

**Comment by Mathias-g** [comment]
@mjobin-mdsol It looks like messages are timestamped and get a unique message ID as well, so they are probably ordered client side based on the timestamps. That should at least mean ordering among multiple devices of the same user should be the same I think? As far as other users go, that is a different matter of course.

**Comment by nihilist001** [comment]
Moving this here as i didn't see there was a duplicate. Though honestly this issue lacks details. here's how i propose this problem is tackled:

Feature

there should be a reliable (and easy) way of using one simplex account from multiple devices:

    master device with the account (on desktop for example)
    secondary device (on mobile for example): you should be able to use the master account on your desktop from your mobile, and without the "same network" requirement, totally unecessary, unless if i'm missing something
    Just allow to do that over any simplex relay server, with a validation mechanism to be done from the master device to authorize (or decline) the synchronisation.

idea:

    on the master device account you create a "use this account from another device" QR code / link
    on that secondary device when you click "create account" you should have the option to "use an account from another device"
    scan that QR code, and then you need to confirm the device sync request from the master account. (just make the verification process include the already existing "verify security code" to make sure the request is valid)
    and from there you should be able to use the account that is on your desktop, from your mobile device.
    there should be a "revoke access" or "manage devices" setting somewhere aswell, in case if you need to edit this later down the road

this is starting to look stupid to require N accounts when you have N devices "user, user-phone, user-desktop, user-laptop" it's a major drawback since you need to check all devices for new private messages or requiring to make a group chat just to make sure all of your devices recieve the private messages.

i think this one should be explicitely put on the roadmap, because look at everyone coming from other chat apps that are expecting this feature to be there, it might deter them from switching to simplex

since simplex supports .onion traffic i don't see the issue. locally just make the user connect to the local SOCKS5 proxy on 127.0.0.1 on port 9050 and then make it publish a hidden service (ndzqjqoziufj...onion:7070) and there will be a permanent way of accessing the master account without requiring any port forwarding of any kind.

**Comment by mjobin-mdsol** [comment]
I agree with Kevin, ...

Even if one device is marked as primary, I definitely want a secondary device to be able to become primary should I lose access to the primary device. Volcano drop involved or not.

**Comment by erebion** [comment]
If you seriously want people to adopt this and yet another system, make it easy to use on multiple devices.

I'm interested in Simplex, but will not adopt anything that limits my usage to one device, as that'd be a huge source of pain in my daily life.

I've tried it out, but upon noticing this limitation, I stopped using it at least until this changes.

**Comment by k-kozika** [comment]
I am a little confused.
1. Is trying to make it possible to the `Use from desktop` feature can be used from outside the local network?
2. Or is this trying to allow messages to be sent and received as the same profile without any connection at all between each device if they are synchronized in some way beforehand?

**Comment by CyberPingU** [comment]
To be honest I'm in trouble even trying to connect the phone to che in-LAN arch linux PC: it says no route to host in the json error file, but I assure I'm on the very same LAN and each machine see's other's layer 2 (broacast) and 3 (direct) packets.

**Comment by OTonGitHub** [comment]
I think the UI choices for "Use from desktop" and "Migrate from device" aren't clear or easy to understand. There's no clear indication that these options rely on your local network. Also, it can trick users into thinking it's like syncing across multiple devices because most apps allow using the same profile everywhere. But with SimpleX, when you connect your mobile device to desktop, your mobile session completely takes over, meaning you're stuck using just one device at a time. It's basically just a temporary switch, not actual syncing. Plus, things like the port number aren't even labeled clearly—you don't even see the word "Port" anywhere. I'm not criticizing the overall design, just the UI part. It would really help if SimpleX made it clear that you're limited to using only one device at a time, maybe with a popup explaining this when you first try using these features or during app installation.

**Comment by azalty** [comment]
Idk if it's possible for SimpleX, but something like Signal's implementation of this could be nice I guess

**Comment by egrieco** [comment]
> Idk if it's possible for SimpleX, but something like Signal's implementation of this could be nice I guess

PLEASE do not follow Signal's lead here. Their implementation is not good. Devices routinely "unlink" from the account if not used for a week or two.

**Comment by Nixo1234** [comment]
I think that the multi-device synchronization is the most important feature to add now to Simplex.
This work with Whatsapp... When I write a message, either from my smartphone, either from my laptop, the chat is synchronized on all device.
I don't want to talk here about how to implement, just about ability to have this feature.
This feature is very important because I don't like to write message on smartphone and sometime when I am using my laptop, it is so comfortable to use a real keyboard.
So, we don't care how it work, even if the implementation of signal, whatsapp or either are bad, we need this feature.
Event if it can work only with 2 or 3 devices, this will be enough for 99% of users.

Please, make a hack, even if it is not perfect at the moment... For instance, each device can have a pair of hidden contacts.
For instance, I have 2 device : A and B with 2 accounts aA on A and aB on B.
Make an "hidden pair" of contact between the accounts aA and AB. Then, instead of sending normal messages, you wrap a synchronization protocol which copy the database contained in A to B and in B to A.
You already have an export and import tool for database. So, you can make a diff analysis of all message in the database from A and B, wrap it  on a sync protocol stacked over the messaging channel between "aA and AB".
This is quick to develop, I think... Not so perfect...

Of course, a proper implementation would need a multiplexing and copy of messages directly on the level of relays... (for instance with should need only one account and 2 sync device.)
But please, just make a temporary hack... It just need some weeks or work... Then, you will be able to upgrade the protocol and make a proper implementation in the future...

We need this only for text messages and photo at the moment... not for voice or video call...
It will not need a lot of resources on server...

**Comment by azalty** [comment]
> PLEASE do not follow Signal's lead here. Their implementation is not good. Devices routinely "unlink" from the account if not used for a week or two.

From my experience this didn't happen

**Comment by egrieco** [comment]
> From my experience this didn't happen

My sample size isn't that much better than one, but it has happened both on my macOS machine previously, and on my current Framework running NixOS. Most of the people I talk to frequently on Signal (at least a dozen) have had it disconnect as well.

At least Signal decided to improve the UX recently. You can now restore current messages when linking a new device, so re-linking is not quite as disruptive as it used to be.

**Comment by egrieco** [comment]
For a positive example, I'd recommend looking into the work of [Ink & Switch](https://www.inkandswitch.com/). The talk [Safe in the Keyhive: Local-first access control with E2EE and capabilities by Brooklyn Zelenka](https://www.youtube.com/watch?v=iLp2xBMud10) covers some related issues. As she mentions, other researchers at Ink & Switch are working on related issues and have also given talks at [Local-First Conf 2025](https://www.localfirstconf.com/).

**Comment by azazar** [comment]
The concept of a "master" device introduces unnecessary hierarchy and complexity. It creates a single point of failure and undermines the decentralization already built into SimpleX. A better approach would treat all devices as equal peers, syncing the same profile across them.

This would be more resilient and user-friendly—especially in scenarios involving device loss or replacement. Requiring a specific "master" device to be online just to keep others functional isn’t sustainable.

What’s needed is true multi-device support: every device operating independently while staying in sync. Like Telegram—but without being Telegram.

**Comment by DNAKentang** [comment]
> The concept of a "master" device introduces unnecessary hierarchy and complexity. It creates a single point of failure and undermines the decentralization already built into SimpleX. A better approach would treat all devices as equal peers, syncing the same profile across them.
> 
> This would be more resilient and user-friendly—especially in scenarios involving device loss or replacement. Requiring a specific "master" device to be online just to keep others functional isn’t sustainable.
> 
> What’s needed is true multi-device support: every device operating independently while staying in sync. Like Telegram—but without being Telegram.

i think no
if we had this feature + make it secure with some sort of key/pass which only the owner that know it
we can make it wayout if something bad happen
like our devices stolen
what we need to do?

make new simplex account? and lost other contact if those phone the ones and only ways of store the data?
so if we had 2 devices, one always on home
and one always carried anywhere

if one of them got stole
we can revoke the stolen ones, with some type of authentication. then make the current ones as main ones.
dunno which u mean it's "single point of failure"
the way it's only 1 and hard to move around is the "single point of failure"
if your phone get stolen, it's fucked up... nothing u can do to make it comeback other than make new ones

**Comment by Nixo1234** [comment]
> The concept of a "master" device introduces unnecessary hierarchy and complexity. 

Hum... there is not "master" device. All devices should be managed by the same way.

> we can revoke the stolen ones, with some type of authentication. then make the current ones as main ones.

Right, we need to revoke stolen device.

We don't know which device will be stolen. It can be computer or smartphone. The revocation should be protected by a special password which is not saved on device. That why, we can not have "a master device".

Revocation is however complicated because it need a signed revocation server like in SSL/TLS.

What about having a pair of RSA keys (private and public) on each device ?

The private key would be encrypted with a password (AES 256 for instance) and stored in each device.

When a device is added to the group of devices, it shares his public key to all other devices.
For instance, if I have 3 devices, A, B and C, I would have :

  On A :    public key of A, B, C   ; aes256 ( private key of A)
  On B :    public key of A, B, C   ; aes256 ( private key of B)
  On C :    public key of A, B, C   ; aes256 ( private key of C)

If the device B is stolen, then, from the device A or the device C, I can revoke B.
To do this, it enter the password on A or C (of course, A and C do not have mandatory the same AES password).

Then, the device A send a signed message to revoke B.
  => this message is forwarded to C.
  => this message is forwarded to all contacts.
  
  By the way, all contacts will be notified that B has been revoked.

**Comment by DNAKentang** [comment]
> Hum... there is not "master" device. All devices should be managed by the same way.

correct
welp its why we always keep evolving and understand something more better than what it's in past, atleast i can justified those as "our understanding evolving" which on past only locked on "master-slave" system but foundout it's more than that

> Right, we need to revoke stolen device.
> 
> We don't know which device will be stolen. It can be computer or smartphone. The revocation should be protected by a special password which is not saved on device. That why, we can not have "a master device".
> 
> Revocation is however complicated because it need a signed revocation server like in SSL/TLS.
> 
> What about having a pair of RSA keys (private and public) on each device ?
> 
> The private key would be encrypted with a password (AES 256 for instance) and stored in each device.
> 
> When a device is added to the group of devices, it shares his public key to all other devices. For instance, if I have 3 devices, A, B and C, I would have :
> 
> On A : public key of A, B, C ; aes256 ( private key of A) On B : public key of A, B, C ; aes256 ( private key of B) On C : public key of A, B, C ; aes256 ( private key of C)
> 
> If the device B is stolen, then, from the device A or the device C, I can revoke B. To do this, it enter the password on A or C (of course, A and C do not have mandatory the same AES password).
> 
> Then, the device A send a signed message to revoke B. => this message is forwarded to C. => this message is forwarded to all contacts.
> 
> By the way, all contacts will be notified that B has been revoked.

yeah that's good enough idea. i mean the current deadlock is if we want to send something on laptop
we should connect it (temp) and then send it
which is pain in the ass and wasting some time

that's why somehow i also support this idea about "Multi-device synchronization" where we can just make it looks like whatsapp where we can connect like 2 devices to be 1. so if one of them received text
and we example isn't on home (using primary or A devices) we can still open it on phone or B devices.

and maybe we can had optional system like....
A devices is on my home, my home is on mars, no one can steal it since i only ones who live on mars.
then i grant the A devices to had ability to revoke any devices

then B devices can only sync without revoke ability since they're phone and can be stolen
something like that (but dunno maybe other people had more amazing idea on future)

we on here talking about the most possible, simple solution on nutshell afterall... maybe dev can see that and try to implement something that make sense more than the other
but still my hard line is same, current Multi-device synchronization which is temp isn't option since it's not that seamless and wasting time if we want to use it....

**Comment by Nixo1234** [comment]
[Sorry, I thought also about something else...]

The solution previously explained seems "relatively" easy to implement. However, what happen if the user own only one device (or if all his devices have been stolen at the same time) ?
Actually, this problem is already existing now because there is no revocation system in SimpleX (or I did not seen... or maybe, there is an issue created... or an issue should be created soon...).
If my device is stolen, the stealer of my phone can continue to chat with my friend, letting them thinking that he is me.
That is, why... hum... we need a kind of revocation server.

For instance, if I have 3 devices, A, B and C, I would have :

  On A :    public key of A, B, C   ; aes256 ( private key of A)
  On B :    public key of A, B, C   ; aes256 ( private key of B)
  On C :    public key of A, B, C   ; aes256 ( private key of C)
  On REVOCATION server :  public key Z   ; aes256 ( private key of Z)

The RSA pair Z has been created by me, on one of my devices and has been sent to the revocation server to be stored there...
It is not stored on device, or it can be of clone of RSA keys of A,B,C...  It is duplicated on the revocation server...
All my contact have received the public key of Z directly from my device  (to avoid a corruption on the SimpleX server).

Then, on the server of SimpleX chat, we need a way to send a message. The user enter his password, the private key Z is unlocked...
A revocation message is sent by the SimpleX server to all my contacts...  [Ah ah... and this is complicated... maybe, we need to publish a list of revocation , like in SSL/TLS revocation server]

By the way, a more complicated revocation server could be added, for the case of all devices have been stolen.

**Comment by azazar** [comment]
> The solution previously explained seems "relatively" easy to implement. However, what happen if the user own only one device (or if all his devices have been stolen at the same time) ? Actually, this problem is already existing now because there is no revocation system in SimpleX (or I did not seen... or maybe, there is an issue created... or an issue should be created soon...). If my device is stolen, the stealer of my phone can continue to chat with my friend, letting them thinking that he is me. That is, why... hum... we need a kind of revocation server.

Isn't that problem out of scope for this issue?

**Comment by Nixo1234** [comment]
> A devices is on my home, my home is on mars, no one can steal it since i only ones who live on mars.

Actually, all device can be stolen... Even on Mars... We do not live in a perfect democracy and sometimes, even police of corrupted judge can steal a device on mars ;-)

That why, a device should not has more access rights than other devices.

The proposed solution allow each device to revoke another (if we know the password). However, we need to keep at least one device away from stealer.

> Isn't that problem out of scope for this issue?

Yes... right... but with this solution, we can revoke devices, only if we still own one !


>  On REVOCATION server : public key Z ; aes256 ( private key of Z)

NB : By the way, I would say that the revocation server should allow to send a message to all my contacts (or publish a message on a revocation list)   : The police or a robber stolen all my devices (computer, smartphone). All my device A,B,C have be revoked. Here is my new simpleX address : xxxx.
By the way, the revocation server can be useful... But, yes, a simple solution for people having at least 2 devices can be quickly implemented. And the problem of revocation exists at the moment... It should be another issue.

**Comment by DNAKentang** [comment]
> > A devices is on my home, my home is on mars, no one can steal it since i only ones who live on mars.
> 
> Actually, all device can be stolen... Even on Mars... We do not live in a perfect democracy and sometimes, even police of corrupted judge can steal a device on mars ;-)
> 
> That why, a device should not has more access rights than other devices.
> 
> The proposed solution allow each device to revoke another (if we know the password). However, we need to keep at least one device away from stealer.

Agree, but somehow for decent people who had enough devices but not the spare ones. it's good to had backup plan protected by password. so we can revoke, remove, or add other else and vice versa when stolen....

perfect world created from corrupt world, when the perfect system take over. the corrupt ones will had no place to go other than dead or being left alone

**Comment by epoberezkin** [comment]
I've added some comments on multi-device support to FAQ: https://github.com/simplex-chat/simplex-chat/pull/6063

The permanent link is here: https://simplex.chat/faq/#why-cant-i-use-the-same-profile-on-different-devices

All the ideas presented above are workable, but it's incorrect to say that either of them is simple - it's an oversimplification, and ignores all the potentials for both securtiy and UX degradation, and other regression risks.

It would be a major project that we do not plan to undertake this year.

**Comment by epoberezkin** [comment]
Threema approach linked above (https://threema.com/en/blog/md-architectural-overview), while preserves e2e encryption security, substantially increases the required trust to the server (what they call mediator server) and in case of SimpleX network this approach would expose to this server communication graph of the users. It is effectively a "thin" client approach in the FAQ section I linked. While it may be possible to have thin client approach while protecting connection graph, I don't see such a possibility straight away. Holding individual double-ratchet-states on the servers would preserve connection graph privacy, as the servers that hold ratchets for concurrent access would be separate from the messaging servers (and they clients would be able to choose a different network operator for this storage), but the downsides of this approach are also explained in the app.

Overall, I believe that solving multi-device is critically important for mass market adoption, but so are other things:
- reliable and scalable groups and channels,
- reliable message reception, which would require adding delivery path redundancy,
- optional memorizable user addresses, for users who want them, that still protect user connection graph,
- etc.

There is a chance that we will start implementing multi-device in 2026, but it's not a hard promise. Effective designs take time to emerge and to evolve, and rushing the solution without having a robust design would be a mistake.

**Comment by azazar** [comment]
Why not use other nodes for revocation data storage instead of servers? Or maybe nodes that are one's contacts(to ensure these nodes legitimacy)?

**Comment by epoberezkin** [comment]
> Why not use other nodes for revocation data storage instead of servers? Or maybe nodes that are one's contacts(to ensure these nodes legitimacy)?

Sorry, I don't quite understand what you suggest.

**Comment by epoberezkin** [comment]
https://github.com/simplex-chat/simplexmq/pull/1588/files - this RFC explores the idea of server-side storage of ratchet state. 

Seems promising, but overall complexity of state synchronisation (that would be the same as with group approach used by Signal) suggests that implementing "link v2" approach may be justified:

- short links/small qr codes using queue data storage.
- when direct local connection fails, negotiate WebRTC data connection between linked devices via SMP protocol (the same queue that is used to store link).
- use zstd compression for API commands and responses to have lower traffic in case of network connection (https://github.com/simplex-chat/simplex-chat/pull/5776 - already implemented)
- allow concurrent usage of the UI
- support iPad as "desktop" for iOS clients (using iPad for android clients is a hard problem, because of different JSON encoding in swift, with swift->kotlin conversation already supported, and kotlin->swift not).

While it would still require that host device is online, the more important problem of the current setup seems to be that people fail to connect via local network, because of firewall settings, or public/office wifi prohibiting direct p2p connections.

**Comment by SKBotNL** [comment]
> While it would still require that host device is online, the more important problem of the current setup seems to be that people fail to connect via local network, because of firewall settings, or public/office wifi prohibiting direct p2p connections.

That is definitely true, however even while using a network where all of that does go right one of my pain points is that it seems like I always have to open up the "Linked mobiles" and "Use from desktop" pages on my desktop and phone respectively for the clients to see each other. This is fine for the initial connection but having to do that every time they disconnect becomes very cumbersome.
A feature that makes the desktop and phone connect automatically in the background would massively improve the user experience IMO.

**Comment by alopatindev** [comment]
> This is fine for the initial connection but having to do that every time they disconnect becomes very cumbersome.

I agree.

When the phone locks, the desktop app disconnects. That happens even with "SimpleX Lock" off in the mobile app. It's hard to finish typing any long message when you get interrupted so often for no reason and need to make so many actions to restore.

**Comment by OTonGitHub** [comment]
I feel like it should be clearly stated that this is not the intended design for SimpleX on their website, during installation and on the the appstore.

**Comment by jerry-harm** [comment]
for now, can I at least use phone to 'remote control' my computer or cli? which should be more reasonable in case of the battery useage.

**Comment by SoongVilda** [comment]
> > While it would still require that host device is online, the more important problem of the current setup seems to be that people fail to connect via local network, because of firewall settings, or public/office wifi prohibiting direct p2p connections.
> 
> That is definitely true, however even while using a network where all of that does go right one of my pain points is that it seems like I always have to open up the "Linked mobiles" and "Use from desktop" pages on my desktop and phone respectively for the clients to see each other. This is fine for the initial connection but having to do that every time they disconnect becomes very cumbersome. A feature that makes the desktop and phone connect automatically in the background would massively improve the user experience IMO.

Jami have pretty great solution for it. :)

**Comment by iMonZ** [comment]
So nothing new?

**Comment by dhana-exe** [comment]
seems they stuck somewhere... hopefully it can still continue
this is the last things we need other than battery drainer approach before massively use the simplex for sure....

**Comment by dipping5115** [comment]
For me, security is far more important than convenience, so I agree to maintain the status quo until a good solution is found.

I think the idea of automatic device matching within the same local network is a good one, or here’s another thought: could we bundle multiple accounts together? A bundled account would be treated as a single entity, with a unified nickname (like a group), and adding a user as a friend to a bundled account would be considered creating a group. However, if a message sent from another identity within the bundled accounts is received in that “group,” it would also be treated as sent by the user themselves. This way, even without using the same identity, message synchronization could be achieved, and perhaps it could also avoid underlying security vulnerabilities (roughly speaking...).

By the way, I’m not really clear on how SimpleX software works, and I’m not sure if I’ve explained this clearly. So, if anyone thinks this solution is completely impractical or even a bit ridiculous, I understand. However, I still want to mention that perhaps my approach could spark some inspiration :D

**Comment by dhana-exe** [comment]
> For me, security is far more important than convenience, so I agree to maintain the status quo until a good solution is found.
> 
> I think the idea of automatic device matching within the same local network is a good one, or here’s another thought: could we bundle multiple accounts together? A bundled account would be treated as a single entity, with a unified nickname (like a group), and adding a user as a friend to a bundled account would be considered creating a group. However, if a message sent from another identity within the bundled accounts is received in that “group,” it would also be treated as sent by the user themselves. This way, even without using the same identity, message synchronization could be achieved, and perhaps it could also avoid underlying security vulnerabilities (roughly speaking...).
> 
> By the way, I’m not really clear on how SimpleX software works, and I’m not sure if I’ve explained this clearly. So, if anyone thinks this solution is completely impractical or even a bit ridiculous, I understand. However, I still want to mention that perhaps my approach could spark some inspiration :D

people already proposed that idea above mate, but it rejected.

**Comment by 2Belette** [comment]
I second that having support of multi-devices would increase mass adoption for SimpleX, as it is a complex problem to solve, would be nice to have something small but would increase a lot the usability (for example what I propose here #6553 very simple but doing this 4/5 times a day is cumbersome, specifically because even if it is local and no firewall and no specific timeout, the connection is breaking time to time.
Another aspect is to have /image and /file working on CLI when using the mobile 'Use from desktop' (at least on my side I am unable to send an file from CLI or decrypt correctly, it may be a one off but I don't know how to build Desktop on FreeBSD, I only found how to build CLI.

All of that small things would already improve the user experience before having a proper / more complex to develop multi-device support

**Comment by OTonGitHub** [comment]
The SimpleX Website is still not updated to reflect that this is the default behavior of SimpleX. This thread keeps growing because this is a User Experience issue. Okay, app is intended this way, it's  not the norm for communication apps these days, just lay it out on your front page so users are informed and don't end up here?

**Comment by manyfew** [comment]
I'm new to SimpleX and this feature for me is a deal breaker, as is lack of an emoji picker (call me shallow). I also find there are settings all over the place and it's really hard to grasp what's where. Learning curve is very steep and it's not ready for mass adoption. I am a tinkerer so I am happy to put in the effort to work my way through and research the settings but the average users would find this extremely intimidating, and I cannot recommend it to most of the people I regularly message.

**Comment by xa4hf8** [comment]
@SKBotNL 

> **A feature that makes the desktop and phone connect automatically in the background would massively improve the user experience**

It could use ZeroConf for device discovery

___

@OTonGitHub 

> **I feel like it should be clearly stated that this is not the intended design for SimpleX on their website, during installation and on the appstore.**

I think it would be sufficient to display a notice in the profile creation dialog, for example:

⚠️ **important notice about multi-device sync**

_SimpleX does not support using the same profile on multiple devices because no secure profile sync implementations have been developed yet. However, a mobile device profile can be used in the desktop app while both devices are connected to the same LAN, if the switch does not enforce client isolation. If you need to use multiple devices in other situations, please create a new profile for each new device._

https://simplex.chat/faq/#why-cant-i-use-the-same-profile-on-different-devices

**Comment by Theo-Z** [comment]
> I think it would be sufficient to display a notice in the profile creation dialog, for example:
> 
> [...]

I disagree - in the short term something like that might be a good idea but improving usability should be a high priority for simplex IMO. Low usability is a big security concern as a tool that you don't use can't protect you.

Multiple of my friends (who are less aware of/ care less about the dangers of surveillance) are already hesitant of switching to services like signal that are basically identical to big tech services and have no learning curve and comfort tradeoffs. SimpleX lacking important features like multiple device support makes convincing them exponentially harder. So them switching takes months longer or never happens and thus our online communication is not sufficiently protected.

**Comment by OldTechNeck** [comment]
It seems like this could almost be created with some of the primitives already available in the software.  I'm thinking under the covers here based on the functionality I can see in the app.  So this is just an idea:

Invent a new secret group type where members cannot see other member's posts, and anything they type goes to the admins just like the chat with admins does now and admins reply directly to them.  Say this is new group type is called a multi-device group.

So now, I would create a new multi-device group, add all my devices as admins, and send the invite to the group out to my friends.  Each one joins and now anything they type only goes to my devices because of the above rule.  This is half the equation.

The problem is that for the users I add, their's would still be a single device so anything I type as admin would go to them but only on their one device address.  However, if I could make it where a multi-device group could be a member of another multi-device group, then both ways would work.  Instead of inviting my friends at their device address, I would invite them at their multi-device group address that they created.

Fun huh?  If I ever get time to play with bots, I might try out something like the above.  There are too many manual steps but it seems at least almost possible with the function set that exists now :)

**Comment by CherrieTheShifter** [comment]
Yes This

**Comment by manyfew** [comment]
Low usability = Low adoption

I use the "use on desktop" function and it's ok while it works but there are some important friction points:

- At the moment it is unreliable, maybe the app on mobile stops working (I think Android terminates it if it detects it's not being actively used), which causes the connection to break, requiring manual reconnect. Re-establishing the connection requires clicking the "wi-fi" icon in SimpleX for desktop to interrupt the link, then going into the menu, clicking "Linked mobiles", clicking the mobile device, then on the mobile device you have to tap the SimpleX app icon, wait for it to load AGAIN because Android had terminated it, tapping the avatar menu, selecting "Use from desktop" and then nothing happens. I have to repeat this procedure shuffling the order of the steps until it happens to work. Definitely a broken UX. The UX needs to involve just one or two intuitive taps, not this much menu diving, and it needs to work reliably for people to want to use it.
- The connection requires confirmation on the mobile device, which is one extra step. I just want to tap "Use on dekstop" and if the dekstop client is already running it should connect without me having to first select the mobile device I want to display on the desktop, etc. I realise this caters to a scenario where multiple mobile devices may be connected, but 99% of users will not benefit from this, so there should be an easy way to connect a mobile device designated as the "main/favorite device".
- If I want to make a voice call, usually I'll do that on the mobile device, which forces me to go back (back key on the phone), which drops the connection on desktop. THIS is a major usability problem.
- The connection on desktop drops constantly.

I think the biggest UX hurdle is not being able to simultaneously operate one account simultaneously from the mobile and the PC. THIS is what needs to be addressed. I don't mind one device being the "server", but simultaneous operation is a must.

I would also like the ability to "Use from mobile" if I primarily use SImpleX on desktop. All devices should be treated equal IMO. It probably shouldn't be "use from desktop" or "use from mobile", but rather, "mirror on other device".

**Comment by azazar** [comment]
Ability to simultaneously operate one account simultaneously from the mobile and the PC is an actual Telegram's killer feature. You can't switch to anything after you had it.

**Comment by manyfew** [comment]
> Ability to simultaneously operate one account simultaneously from the mobile and the PC is an actual Telegram's killer feature. You can't switch to anything after you had it.

It's definitely the biggest friction point and the #1 complaint frmo most people I've recommended it to.
